parsers/US_IPC.py

```diff
@@ -0,0 +1,145 @@
+#!usr/bin/env python3
+
+"""Parser for the Idaho Power Comapny area of the United States."""
+
+from dateutil import parser, tz
+from itertools import groupby
+from logging import getLogger
+import requests
+
+
+# NOTE No pumped storage yet but future ideas can be found at the following url.
+# https://docs.idahopower.com/pdfs/AboutUs/PlanningForFuture/irp/IRP.pdf
+
+PRODUCTION_URL = 'https://api.idahopower.com/Energy/Api/V1/GenerationAndDemand/Subset'
+
+GENERATION_MAPPING = {'Non-Utility Geothermal': 'geothermal',
+                      'Hydro': 'hydro',
+                      'Coal': 'coal',
+                      'Diesel': 'oil',
+                      'PURPA/Non-Utility Wind': 'wind',
+                      'Natural Gas': 'gas',
+                      'PURPA/Non-Utility Solar': 'solar',
+                      'PURPA Other': 'biomass'
+                      }
+
+
+def get_data(session=None):
+    """Returns a list of dictionaries."""
+
+    s = session or requests.Session()
+
+    req = requests.get(PRODUCTION_URL)
+    json_data = req.json()
```

Comment:
Hm... weird behaviour of the API. Sometimes I get parser failures:

```
Traceback (most recent call last):
  File "US_IPC.py", line 143, in <module>
    print(fetch_production())
  File "US_IPC.py", line 121, in fetch_production
    raw_data = get_data(session=session)
  File "US_IPC.py", line 34, in get_data
    json_data = req.json()  
  File "/home/jarek/projects/electricitymap/venv3/lib/python3.4/site-packages/requests/models.py", line 866, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3.4/json/__init__.py", line 318, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.4/json/decoder.py", line 343, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.4/json/decoder.py", line 361, in raw_decode
    raise ValueError(errmsg("Expecting value", s, err.value)) from None
ValueError: Expecting value: line 1 column 11592 (char 11591)
```

where the JSON around that point looks like (linebreaks inserted):

```
...
{"name":"Coal","data":157.000000000000,"datetime":"2018-03-28T12:00:00"},
{"name":"Hydro","data":1318.000000000000,"datetime":"2018-03-28T12:00:00"},
{"name":"Net Purchases","data":0.000000000000,"datetime":"2018-03-28T12:00:00"},
{"name":"Inadvertent","data":***************,"datetime":"2018-03-28T13:00:00"},
{"name":"Load","data":1455.000000000000,"datetime":"2018-03-28T13:00:00"},
{"name":"Non-Utility Geothermal","data":32.000000000000,"datetime":"2018-03-28T13:00:00"},
{"name":"PURPA Other","data":39.000000000000,"datetime":"2018-03-28T13:00:00"},
...
```

That is quite clearly an API problem producing invalid JSON. I don't know how often it happens - right now it's about 30% of the time for me, but that datapoint is almost 3 days old and I don't know how often they insert such invalid things, or why it's only returned some of the time. Maybe let's merge and see in logger how often it causes problems?

Reply:
Agreed. I saw this a few times but the weird thing was that when I ran the parser again immediately afterwards the json was fine.